src/test/scala/tech/sourced/gemini/CassandraSparkSpec.scala

```diff
@@ -107,6 +149,96 @@ class CassandraSparkSpec extends FlatSpec
     sha1.size shouldEqual 2
   }
 
+  def bblfshMock(uast: Node): ProtocolService = {
+    class BblfshServerMock extends ProtocolService {
+      override def parse(request: ParseRequest): Future[ParseResponse] = {
+        Future.successful(ParseResponse(uast = Some(uast)))
+      }
+
+      override def nativeParse(request: NativeParseRequest): Future[NativeParseResponse] = {
+        Future.successful(NativeParseResponse())
+      }
+
+      override def version(request: VersionRequest): Future[VersionResponse] = {
+        Future.successful(VersionResponse())
+      }
+    }
+
+    new BblfshServerMock
+  }
+
+  def feMock(features: Seq[Feature]): FeatureExtractor = {
+    class FEServerMock extends FeatureExtractor {
+      override def identifiers(request: IdentifiersRequest): Future[FeaturesReply] = {
+        Future.successful(FeaturesReply(features=features))
+      }
+
+      override def literals(request: LiteralsRequest): Future[FeaturesReply] = {
+        Future.successful(FeaturesReply())
+      }
+
+      override def uast2Seq(request: Uast2seqRequest): Future[FeaturesReply] = {
+        Future.successful(FeaturesReply())
+      }
+    }
+
+    new FEServerMock
+  }
+
+  def readFeaturesFromFile(path: String): Seq[Feature] = {
+    val file = new File(path)
+    val byteArray = Files.readAllBytes(file.toPath)
+    val jsonData = JSON.parseFull(new String(byteArray))
+
+    jsonData match {
+      case Some(m: Map[_, _]) => {
+        val featuresMap = m.asInstanceOf[Map[String, Double]]
+        val iter = featuresMap.map { case(key, value) =>
+          Feature(name = key, weight = value.toInt)
+        }
+        iter.toSeq
+      }
+      case _ => throw new Exception("incorrect json")
+    }
+  }
+
+  "Query for similar files" should "return 1 file" in {
```

Comment:
The test looks awesome to me:
 - it uses mocks for FE and Bblfsh
 - it loads to DB from .json fixtures for two hashed files, one of which is src/test/resources/models.py
 - FE extractor mock always return features, extracted for this file
 - and then it checks that 1 similar file is returns

👍